### PR TITLE
Add support for exempting i18n/translation commits from Redmine validation

### LINKED ISSUE #199: Add support for exempting i18n/translation commits from Redmine validation
## Problem

Currently, all commits in repositories with `redmine_required: true` must follow the format `fixes #1234 - description` or `refs #1234 - description` to pass validation. However, translation/internationalization commits often don't have specific Redmine issues associated with them, causing these legitimate commits to fail validation.

This issue originated from discussion in https://github.com/theforeman/theforeman-rel-eng/pull/519#discussion_r2313542525 where we need to handle locale/translation updates that may not have specific Redmine issues but still need to pass validation.

## Proposed Solution

Add support for `exempt_patterns` configuration that allows certain commit message patterns to bypass Redmine validation requirements while maintaining validation for regular commits.

### Features:
- **Global default patterns** applied to all repositories (for common i18n patterns)
- **Repository-specific override** capability for special cases
- **Regex pattern matching** for flexible commit message formats
- **Backward compatible** - no breaking changes to existing configuration

### Example Configuration:
```yaml
# Global patterns applied to ALL repositories (unless overridden)
global_config:
  exempt_patterns:
    - "^i18n - "
    - "^Localization - "
    - "^Update translations"
    - "^Automatic locale update"

# Repository uses global patterns automatically
theforeman/foreman:
  redmine: foreman
  redmine_required: true

# Repository can override with custom patterns
theforeman/special-repo:
  redmine: special
  redmine_required: true
  exempt_patterns:
    - "^Bot update - "
    - "^Automated sync - "
```

### Example Exempt Commits:
- ✅ `i18n - Update Japanese translations`
- ✅ `Localization - Fix German strings`
- ✅ `Update translations for 3.5 release`
- ✅ `Automatic locale update from Crowdin`

Regular commits would still require Redmine references:
- ❌ `Add new user management feature` (needs `fixes #1234 -`)
- ✅ `fixes #1234 - Add new user management feature`

This solution addresses the core issue while preserving the validation requirements that ensure proper issue tracking for development work.

### PR DESCRIPTION
**Pull request by ogajduse**
## Summary
* Add exempt_patterns configuration to allow specific commit message patterns to bypass Redmine validation requirements
* Implement global default patterns applied to all repositories with repository-specific override capability
* Add regex pattern matching for flexible commit message formats
* Maintain backward compatibility with existing configuration

## Background
Resolves the issue discussed in https://github.com/theforeman/theforeman-rel-eng/pull/519#discussion_r2313542525 where translation/internationalization commits often don't have specific Redmine issues associated with them, causing legitimate commits to fail validation.

## Changes
* Added `exempt_patterns` field to Config dataclass
* Updated configuration loading to support global_config section with repository override behavior
* Implemented `is_commit_exempt()` function with regex pattern matching
* Modified commit validation in `get_issues_from_pr()` to check exemptions before marking commits invalid
* Added global i18n patterns to repos.yaml for common translation workflows
* Fixed edge case with empty commit messages in subject property

## Global Patterns Added
* `"^i18n - "` - Matches "i18n - Update Japanese translations"
* `"^Localization - "` - Matches "Localization - Fix German strings"
* `"^Update translations"` - Matches "Update translations for 3.5 release"
* `"^Automatic locale update"` - Matches "Automatic locale update from Crowdin"

## Configuration Examples
```yaml
# Repository uses global patterns (no local exempt_patterns defined)
theforeman/foreman:
  redmine: foreman
  redmine_required: true
  # Gets global patterns automatically

# Repository completely overrides global patterns  
theforeman/special-repo:
  redmine: special
  redmine_required: true
  exempt_patterns:
    - "^Bot update - "
  # Gets only these patterns (global patterns replaced)
```

## Test plan
- [x] Manual testing of exempt pattern matching functionality
- [x] Verification of configuration loading with global and repository-specific patterns
- [x] Testing edge cases (empty commits, no patterns)
- [x] Validation that existing behavior remains unchanged for repos without exempt patterns

🤖 Generated with [Claude Code](https://claude.ai/code)

### CODE REVIEW

File: prprocessor/__main__.py [hunk 1]
```diff
@@ -94,6 +110,15 @@ class Config:
 logger = logging.getLogger('prprocessor')  # pylint: disable=invalid-name
 
 
+def is_commit_exempt(commit: Commit, exempt_patterns: list) -> bool:
+    """Check if commit subject matches any exemption pattern."""
+    import re
```
Comment (ekohl) [inline prprocessor/__main__.py]: Can this be a top level import?

File: prprocessor/__main__.py [hunk 2]
```diff
@@ -94,6 +110,15 @@ class Config:
...
+    for pattern in exempt_patterns:
+        if re.search(pattern, commit.subject, re.IGNORECASE):
+            return True
+    return False
```
Comment (ekohl) [inline prprocessor/__main__.py]: Maybe a bit of code golfing
```suggestion
    return any(re.search(pattern, commit.subject, re.IGNORECASE) for pattern in exempt_patterns)
```

File: prprocessor/__main__.py [hunk 3]
```diff
@@ -214,7 +239,8 @@ async def get_issues_from_pr(pull_request: Mapping) -> tuple[IssueValidation, Co
         issue_ids.update(commit.fixes)
         issue_ids.update(commit.refs)
         if config.required and not commit.fixes and not commit.refs:
-            invalid_commits.append(commit)
+            if not is_commit_exempt(commit, config.exempt_patterns):
```
Comment (ekohl) [inline prprocessor/__main__.py]: Any particular reason you didn't add this to the existing `if` statement? Would the line become too long?

File: prprocessor/config/repos.yaml [hunk 4]
```diff
@@ -4,7 +4,17 @@
...
+#                    Redmine validation (useful for i18n, automated updates, etc.)
 ---
+
+# Global default patterns applied to ALL repositories (unless overridden)
+global_config:
+  exempt_patterns:
+    - "^i18n - "
+    - "^Localization - "
+    - "^Update translations"
+    - "^Automatic locale update"
```
Comment (ekohl) [inline prprocessor/config/repos.yaml]: I don't like the many variations for this. In our code base we only have 2 and let's keep it that way
```suggestion
    - "^i18n - (extracting new, )?pulling from tx$"
```

Sources:
* https://github.com/theforeman/foreman/blob/ee1afb939ebfea220177b5066c4bb9744721cdfe/locale/Makefile#L62
* https://github.com/theforeman/foreman_plugin_template/blob/f92b585db9ddd4d8fc9c2bc3209162962157a67c/locale/Makefile#L70
* https://github.com/theforeman/hammer-cli/blob/737ec089d22d01aa537764dc9b6be0f1b8718a64/lib/hammer_cli/i18n/find_task.rb#L92

**Review by ekohl — CHANGES_REQUESTED** [review]
I think the whole config is a bit excessive. There should only be a single pattern and I'd be inclined to hardcode it. Then you can use `re.compile` and reuse the compiled pattern. You can also simplify the code a lot by making it a simple `is_exempt` method on the `Commit` class. That keeps rest of the logic easier as well.

File: prprocessor/__main__.py (hunk 2 above)
Comment (ogajduse) [reply prprocessor/__main__.py]: I was thinking about that too.

File: prprocessor/__main__.py (hunk 3 above)
Comment (ogajduse) [reply prprocessor/__main__.py]: If we had some specific (I do not exactly remember which one) flake ruleset enabled here, it would point out the very same thing. Will fix it.

File: prprocessor/config/repos.yaml (hunk 4 above)
Comment (ogajduse) [reply prprocessor/config/repos.yaml]: Right, sort of wanted to have a discussion about this - if the `i18n` one is the only one we want to have in at this moment. 

I'll leave only that one there then.

**Comment by ogajduse** [comment]
I've tried to address all your comments. Could you please re-review @ekohl?

File: prprocessor/__main__.py [hunk 5]
```diff
@@ -26,6 +26,7 @@
     re.IGNORECASE,
 )
 COMMIT_ISSUES_REGEX = re.compile(r'#(\d+)')
+EXEMPT_COMMIT_REGEX = re.compile(r'^i18n\s*-', re.IGNORECASE)
```
Comment (ekohl) [inline prprocessor/__main__.py]: I think we can be more specific here to keep them standardized. Since they're typically generated by scripts that should be fine.
```suggestion
EXEMPT_COMMIT_REGEX = re.compile(r'^i18n - (extracting new, )?pulling from tx$', re.IGNORECASE)
```

**Review by ekohl — COMMENTED** [review]
Thanks! I think this is a lot simpler to understand (and thus maintain).

File: prprocessor/__main__.py (hunk 5 above)
Comment (ogajduse) [reply prprocessor/__main__.py]: Changed.
